control-plane-operator/controllers/hostedcontrolplane/routecm/reconcile.go

```diff
@@ -0,0 +1,239 @@
+		return fmt.Errorf("failed to get openshift controller manager config: %w", err)
+	}
+
+	configBytes, ok := config.Data[configKey]
+	if !ok {
+		return fmt.Errorf("openshift controller manager configuration is not expected to be empty")
+	}
+	configHash := util.ComputeHash(configBytes)
+
+	maxSurge := intstr.FromInt(0)
+	maxUnavailable := intstr.FromInt(1)
+	deployment.Spec.Strategy.Type = appsv1.RollingUpdateDeploymentStrategyType
+	deployment.Spec.Strategy.RollingUpdate = &appsv1.RollingUpdateDeployment{
+		MaxSurge:       &maxSurge,
+		MaxUnavailable: &maxUnavailable,
+	}
+	if deployment.Spec.Selector == nil {
+		deployment.Spec.Selector = &metav1.LabelSelector{
+			MatchLabels: openShiftRouteControllerManagerLabels(),
+		}
+	}
+	deployment.Spec.Template.ObjectMeta.Labels = openShiftRouteControllerManagerLabels()
+	deployment.Spec.Template.ObjectMeta.Annotations = map[string]string{
+		configHashAnnotation: configHash,
+	}
+	deployment.Spec.Template.Spec.AutomountServiceAccountToken = ptr.To(false)
+	deployment.Spec.Template.Spec.Containers = []corev1.Container{
+		util.BuildContainer(routeOCMContainerMain(), buildRouteOCMContainerMain(image)),
+	}
+
+	return nil
+}
+
+var (
+	volumes = component.Volumes{
+		routeOCMVolumeConfig().Name: component.Volume{
+			Source: component.ConfigMapVolumeSource(ConfigMapName),
+			Mounts: map[string]string{
+				routeOCMContainerMain().Name: "/etc/kubernetes/config",
+			},
+		},
+		routeOCMVolumeServingCert().Name: component.Volume{
+			Source: component.SecretVolumeSource(manifests.OpenShiftRouteControllerManagerCertSecret("").Name),
+			Mounts: map[string]string{
+				routeOCMContainerMain().Name: "/etc/kubernetes/certs",
+			},
+		},
+		routeOCMVolumeKubeconfig().Name: component.Volume{
+			Source: component.SecretVolumeSource(manifests.KASServiceKubeconfigSecret("").Name),
+			Mounts: map[string]string{
+				routeOCMContainerMain().Name: "/etc/kubernetes/secrets/svc-kubeconfig",
+			},
+		},
+		common.VolumeTotalClientCA().Name: component.Volume{
+			Source: component.ConfigMapVolumeSource(manifests.TotalClientCABundle("").Name),
+			Mounts: map[string]string{
+				routeOCMContainerMain().Name: "/etc/kubernetes/client-ca", // comes from the generic OCM config
+			},
+		},
+	}
+)
+
+// Volumes implements controlplanecomponent.DeploymentReconciler.
+func (r *RouteControllerManagerReconciler) Volumes(cpContext component.ControlPlaneContext) component.Volumes {
+	return volumes
+}
+
+func openShiftRouteControllerManagerLabels() map[string]string {
+	return map[string]string{
+		"app":                         "openshift-route-controller-manager",
+		hyperv1.ControlPlaneComponent: "openshift-route-controller-manager",
+	}
+}
+
+func routeOCMContainerMain() *corev1.Container {
+	return &corev1.Container{
+		Name: "route-controller-manager",
+	}
+}
+
+func buildRouteOCMContainerMain(image string) func(*corev1.Container) {
+	return func(c *corev1.Container) {
```

Comment:
This could be handled with a builder pattern as well

Reply:
could you please elaborate more.

Reply:
Instead of having a generic reconcile function where we replicate some code across components, we could instead use a set of functions like 
container("name").Command("blah").Args("foo","bar").WithEnv(blah)

Reply:
done! let me know if that looks good to you.